### ISSUE TITLE
migrate from pickle to a generic serialization protocol

### ISSUE THREAD

**Issue by indirectlylit** [learningequality/kolibri#6316]
### Observed behavior

Currently we persist information about tasks using pickle, but this has caused us some issues when python versions get mixed. This should not actually happen, but could cause issues in the future...

* important for https://github.com/learningequality/kolibri/issues/6255
* potentially a security risk
* slow performance
* may be problematic for multiprocessing

From @rtibbles 

> I am very open to abandoning pickle altogether - it would place more restrictions on the kind of args and kwargs we pass, but I don't think that's a bad thing.
> 
> I think the only reason it is the current mechanism is that it does get pickled eventually to get passed to the executor pool for execution

### Expected behavior

pickle-free task communication

### User-facing consequences

edge case issues

### Context

0.13.0 beta 7

**Comment by benjaoming** [comment]
There's also some SQLalchemy schema dumping/loading for Kolibri channel import/export that might be of interest. It was already using protocol 2, so not of imminent concern.

* https://github.com/learningequality/kolibri/blob/develop/kolibri/core/content/utils/sqlalchemybridge.py
* https://github.com/learningequality/kolibri/blob/develop/kolibri/core/content/management/commands/generate_schema.py

**Comment by rtibbles** [comment]
I don't count this as serialization - so I was not intending to imply we should changes these - just in task serialization etc.

**Comment by vkWeb** [comment]
Does [`json`](https://docs.python.org/3/library/json.html) from the python's standard library make sense here? json will allow us to encode python object hierarchies to their string representation (which can then be stored in db). 

Tbh, I haven't used pickle and also I'm pretty new to kolibri, so I am not sure if this approach will make sense, but just sharing my thoughts...

**Comment by EliKlein** [comment]
I think I'd like to start looking at this, but I'm also pretty new with Kolibri, so I want to make sure I'm looking in the right place. It should just be the Job class that needs to be serialized right? And then the ORMJob and ScheduledJob classes can have their "obj" column type changed and stuff, to finish it off, basically?

I'll check where else these are all used too, but I just wanna be sure I'm looking at the right thing to start with.

Also, json sounds like it might work, but I think I'm also gonna look for other options to be sure.

**Comment by rtibbles** [comment]
Hi @EliKlein - apologies, I hadn't realized that @vkWeb was not assigned to this issue. He was planning to take this on - but if you want to take a stab at it, he is currently busy with some other tasks, so I'd welcome your contribution.

You seem to have the right idea yes - the main complexity in serialization comes from the fact that we are currently pickling the `result` object and any `exceptions` that get stored on the Job class, and those are not as straight forward to serialize to JSON. This would probably require adding some validation logic to ensure that our result objects only contain JSON serializable information when set on the Job, and that we find a systematic way of serializing any exception objects.

**Comment by EliKlein** [comment]
Hm... Yeah, this does seem complicated, but I'm going to give it a go, at least for now.

**Comment by vkWeb** [comment]
@EliKlein I've a firm grasp of everything Task backend related so if you need any help / have questions, feel free to ask here, I'll be happy to help. Excited to see your work!

**Comment by EliKlein** [comment]
@vkWeb Thanks! I've gone through more, and I think I have a good idea of what's what now, but I do have a few questions still.

First, where are the results of the task actually being used? I see a couple functions that are run as Jobs but all of them return None. Seems like you'd only need to know whether the job is complete or not at that point.

Second, I'm not _super_ up on SQLAlchemy, so I'm not entirely sure I know how I can alter the column to make it store data differently (as a string) but still work the same otherwise. I'm looking at this Hybrid Attributes section of the documentation, and wondering about using that. So, I though I should ask, if I do something like

```python
saved_job = Column(String)

# Using the old name "obj" as a Job object still, storing that internally as _job_object
@hybrid_property
def obj(self):
    # Try to get the internal value, or make it if it doesn't exist yet
    try:
        return self._job_object
    except AttributeError:
        self._job_object = fn_to_decode_json(self.saved_job)
        return self._job_object

@obj.setter
def obj(self, value):
    self.saved_job = fn_to_encode_json(value)
    self._job_object = value
```

does `_job_object` stay attached to the row within the session, so that it doesn't have to decode the object all the time? Does this seem like the right approach to you?

**Comment by rtibbles** [comment]
@EliKlein one slight advantage you have in doing this now, is that we will be wiping the jobs database and recreating it from scratch when people upgrade to the version this will be in, so you don't need to worry about pre-existing state.

You can just remove the old column and add the new serialization column in its place.

**Comment by EliKlein** [comment]
That's good to know, though I'm still wondering about some things.

- I don't see the results currently being used anywhere. Are they actually necessary to have in the first place?
- It's not easy to totally avoid the risk of unpickling untrusted data. The current security problem is just as much in using pickle as it is in using `import_stringified_func`. How much of this issue overall is about security, versus a compatibility/speed concern?

**Comment by rtibbles** [comment]
None of this is really about security - it's only application code that access this layer, and it's insulated from injection attacks via validation and the fact that code has to have been written to be executed in the first place.

It's more that pickle is kind of bloated, and not really necessary (as you note, we're not really using the pickled results anywhere anyway) - also, because we are trying to be very cautious in ensuring Python 2 compatibility, we are still using Python 2 compatible pickling which is even slower and less performant.

Using JSON would make the tasks DB much more dev inspectable, and take away having to rely on pickle for this part of our code.

As nothing is really relying on the pickling behaviour at the moment, this is a good time to enforce a JSONic serialization protocol, as it will just shape what we do in the future!

**Comment by indirectlylit** [comment]
With regard to performance, it might be doing some quick timing measurements of before & after behaviors, and also making some estimates of how often these code paths will be executed for a typical classroom server.

If this is a particularly hot code path we might want to consider an alternative json library. For example, [ujson](https://github.com/ultrajson/ultrajson) is much faster, but it's also written in c so we would need to ensure it falls back gracefully on unsupported architectures and Python versions.

**Comment by rtibbles** [comment]
I think any json encoding decoding is likely to be swamped by our json encoding decoding in our API endpoints, and that has not been a performance bottleneck so far. I think this would be a distraction from the simpler goal of just removing pickle.

**Comment by indirectlylit** [comment]
good point. There is a [DRF ujson layer](https://github.com/Amertz08/drf_ujson2) we could look at but this is indeed a distraction. Opened https://github.com/learningequality/kolibri/issues/8427

**Comment by EliKlein** [comment]
This is not going super well. I have the serialization set up properly for simple cases, but it looks like there's trouble with some tasks already using arguments which aren't (normally) serializable to JSON-- Namely, the EventProxy in test_job_running.

I had an idea on how to make it a much smaller problem (or even not a problem at all) but if it's possible, I can't figure it out.

Not sure if I should pursue that idea further, versus trying to force things into being serializable?

There also might be a third option, but I need a working pytest to know if that's a good idea, and pytest is breaking in just the wrong way for me.

**Comment by rtibbles** [comment]
The only important data from the EventProxy is the internal event_id property.

I would suggest updating the tests so that we just pass in the `event_id` of the `EventProxy` as the argument, then update `EventProxy` so that it can take an optional `event_id` argument on initialization.

Within the jobs that are using the EventProxy, you would then initialize an EventProxy with the passed in event_id argument.

This should allow you to just use JSON serialization with minimal changes to the tests.

**Comment by EliKlein** [comment]
OK that was pretty helpful. I've gotten through as much of getting tests to pass as I can, but there's some odd problems with my computer running pytest (I mentioned this before). I just can't run all of the tests at once for some reason, and I'm sure if I tried to find the tests that it's not getting to, I'd miss something.

I'm going to go touch up what I've changed a little bit, for now... and I would ordinarily do a pull request after this is done, but it feels wrong to do that if I can't verify that the code is definitely passing all the tests. Is it still OK to do one if I can't test it myself?

(As far as the problem goes, I'm very much just guessing here, but it might be something to do with my computer specs/RAM, based on the feedback it gives me)

**Comment by EliKlein** [comment]
Decided to do a pull request anyways.

**Comment by rtibbles** [comment]
Thank you! We can review and give feedback in case there's anything else that needs updating. Your contribution is very much appreciated!

**Comment by rtibbles** [comment]
Fixed in #8447
